### ISSUE TITLE
Question: CDI request context usage

### ISSUE THREAD

**Issue by mkouba** [eclipse/microprofile-concurrency#16]
CDI request context is by its nature single-threaded, i.e. users never synchronize `@RequestScoped` beans. What happens if a user sumbits a task to the executor and then continues to work with the current request context:

```java
class MyService {
  // RequestBean is @RequestScoped
  @Inject 
  RequestBean bean;

  @Inject 
  ManagedExecutor executor;
  
  @Inject 
  ThreadContext threadContext;
  
  void doSomething() {
    // myFunction might be executed in parallel
    executor
      .supplyAsync(threadContext.withCurrentContext(myFunction)); 
     // what happens if change the current request context?
     bean.changeMyState();
  } 
```
Without synchronization (on the bean class level) race conditions may occur.

**Comment by FroMage** [comment]
My personal opinion is that this should be up to the user to synchronise concurrent modifications of state across threads. We can only guarantee that contexts are propagated. It is then up to the user to use the proper thread synchronisation or the proper underlying thread model (in vert.x, each request is bound to a single thread).

**Comment by FroMage** [comment]
Well, ATM context is not propagated so it's even easier to break an app now, even without any concurrent access going on ;) It would still be progress and less breakage.

**Comment by mkouba** [comment]
Well, that's probably a different kind of "breaking" ;-). Anyway, I do not question the usefulness of context propagation. I'm trying to point out the difficulties before it's too late...

**Comment by njr-11** [comment]
I think there would be some value in updating the spec doc to emphasize the application's responsibility for using each type of context in a way that is thread safe.  Generally this is something that each provided context type will need to define itself.  We could also consider whether CDI context ought to be listed as one of the types that is propagated by default when the user injects/builds a ManagedExecutor or ThreadContext without specifying a particular set of context types.  And let's get some input from CDI experts such as @Emily-Jiang around whether there is anything more that ought to be done here particular to CDI.

**Comment by Emily-Jiang** [comment]
@mkouba is right. This is indeed an issue. We should discuss all context propogation further in our next hangout.

**Comment by manovotn** [comment]
@njr-11 @Emily-Jiang Since I cannot really make in into those mtg (due to their time), have you came up with anything regarding this question?

What @mkouba brought up is IMO a very good and serious remark. 

Yet another one could be that as long as we are talking actual concurrency and not just async (as in for instance event loop in JS where it is still single threaded), it is going to be error prone. For instance you can attempt to execute something in one thread with current session context, however yet another thread, running prior to the first one, decides to end the session (invalidate it for instance). Then you are in an undefined and unforeseeable state.

**Comment by manovotn** [comment]
To add a bit of explanation to what Martin said: 

* `the set of mapped instances` will actually contain only the beans you have already _used_ within that context. E.g. it is _not_ a set of all request scoped beans, instead it is a set of req. scoped beans you have used so far

* Beans are created lazily - until you actually invoke a method on them, it is just an "empty" proxy object you are getting. E.g. `Instance<AnotherReqBean> anotherBean` does not cause `AnotherReqBean` to be added to the set of mapped instances whereas obtaining a reference and calling `coolPing()` on it does it.

**Comment by njr-11** [comment]
We didn't have anyone representing CDI on at the last MP Concurrency meeting, so the discussion was deferred until this coming Friday.  Based on what is in this issue so far, I'm concerned that the MP Concurrency spec would be overreaching its intended scope in trying to define the meaning of CDI context propagation.  The intent behind the pluggable SPI for thread context propagation was that each individual technology would be able define (and implement) the meaning of context propagation as pertains to it, with MP Concurrency providing a standard SPI and the infrastructure to ensure that providers of the SPI are used to capture, propagate or clear, and restore context as necessary.

Recognizing that some context types might come from EE rather than MicroProfile, and with it being unlikely for the EE context types to get spec updates any time in the near feature, the MP Concurrency spec attempted to cover some of the EE context types as a matter of convenience.  Most of these types (Application, Security, Transaction) have well-defined and obvious meaning with respect to context propagation (it helps that they are all covered in the existing EE Concurrency spec).  CDI however, doesn't and isn't discussed in EE Concurrency, and the further we delve into it, the more clear it is that a definition of context propagation for CDI really belongs in a CDI spec as opposed to being tacked on to the MP Concurrency spec in a way that cannot account for all of the possible scenarios.  Moving the definition to a CDI spec (whether MP or EE) will also have the added benefit that CDI implementations will be more likely to implement it.  Any thoughts on this?

I'll add that I'm still very much in favor of having CDI context be propagated, and would like to see it defined and to see the discussions here continue.  I'm just suggesting that it might better be done tangential to the MP Concurrency spec.

**Comment by FroMage** [comment]
I have to point out however that due to CDI not providing context propagation at the moment, and if MP-Concurrency cannot push CDI to provide context propagation, this limitation prevents reactive applications from using CDI, and in practice reactive applications will be forced to avoid CDI wherever possible, if not abandon it outright, which is not at all what we want.

**Comment by njr-11** [comment]
I agree that is not what we want. The idea would be to push CDI to support context propagation within its own specifications, not to abandon it.

**Comment by Emily-Jiang** [comment]
@FroMage agreed with your points. We need to talk about the CDI context propagation this Friday. Since CDI cannot do it, we have put all the hope on this concurrency spec. If we leave CDI out, I think we will disappoint a lot of developers. @mkouba @manovotn can you join the call so that we can discuss this further. We can at least target some basic scenario and walk our way through it?

**Comment by manovotn** [comment]
Yea, I am planning on showing up this week. Although I have to say current mtg time is a killer for me (17:30 Fri, I am oftentimes not anywhere near PC).

I think the concerns me and @mkouba have originate partly from not seeing the scope of support this spec expects to cover. This makes us presume use cases which may even not be legitimate? Having TCKs drafted (or any other set of user-stories/use-cases) would probably make it a lot clearer.

We will definitely have to draft some solution on Weld-side before this even hits CDI spec - things would move slowly there. Too slowly to wait for it I suppose. In fact, we are even now discussing how to do it in [Weld issue](https://issues.jboss.org/browse/WELD-2497). One thing I am planning on doing tomorrow (or on Fri) is to send an email to cdi-dev and bring their attention to this effort so we may get more opinions on what is actually expected.

**Comment by FroMage** [comment]
If you want an example of typical use-case, look at https://github.com/resteasy/Resteasy/blob/master/resteasy-rxjava2/src/test/java/org/jboss/resteasy/rxjava2/RxResource.java#L62 which is testing that RESTEasy context propagation works:

- Get a `UriInfo` injected via RESTEasy injection with `@Context` (it will be a proxy lazily loaded via the RESTEasy request context, stored in a thread-local)
- Create an RxJava `Single`
- Fullfill the `Single` via a new thread (has no context), which is typical for reactive applications as events are fired by many sources not managed by the container.
- That fullfillment will call the further pipeline operations such as `map` which uses the `UriInfo` proxy
- If that `map` call does not set up the context, the proxy call will fail

This is very typical of reactive applications, where the request will become async, register listeners / pipeline operations, and return. Some external source will then trigger the listeners and awaken the pipeline and the user's callbacks/lambdas will be called, with the expectation/requirement that the callbacks created in the request will actually still have the request scope in order to function.

That's what we're trying to solve. The same is true of course of security/transaction/etc…

**Comment by njr-11** [comment]
@mkouba and @manovotn - would ThreadContext.CDI need to have any prereq dependency on ThreadContext.APPLICATION being established on the thread the same way (for example, both propagated to thread, or both left unchanged on thread)?  I'm trying to determine if the spec really needs the concept of prerequisite context types because it would be a simpler model for the user as well as the MP Concurrency implementation without this concept.  The only reason we initially included it was in anticipating the possibility that CDI might have a need for it.  But if it doesn't, then we would like to remove prereq context types from the initial version of the spec. Please let me know your thoughts on this.

**Comment by manovotn** [comment]
Not sure what ThreadContext.APPLICATION actually brings?

From CDI perspective, the propagator simply needs to have CDI container running to be able to get hold of WeldManager (Weld-enriched version of BeanManager). From there you grab contexts and extract bean instances from then subsequently setting them in a similar way into new thread.
At least that's the summary of my current [PoC for context propagation in Weld](https://github.com/weld/core/pull/1877), still WIP though.

**Comment by njr-11** [comment]
@manovotn thanks for the explanation.  Based on what you said, it's pretty clear that the CDI context doesn't have a dependency on the APPLICATION context type.  ThreadContext.APPLICATION covers the application-related context that is defined in the Java EE Concurrency spec, which is the component/module/application namespace and the thread context class loader for the module.  I'll go ahead and remove the concept of prerequisite context from the spec, because it doesn't have any immediate value  - it can always be added back in the future if the need arises.

**Comment by njr-11** [comment]
Closing out because #138, which addresses this, has been merged
